hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java

```diff
@@ -8938,5 +9084,78 @@ public void stopMonitor() {
     }
    return new ArrayList<>();
   }
+
+  public void setDelayBeforeSTOFlag(long delay){
+    if(isTestingSTO)
+      this.delayBeforeSTOFlag = delay;
+  }
+
+  public void setDelayAfterBuildingTree(long delay){
+    if (isTestingSTO)
+      this.delayAfterBuildingTree = delay;
+  }
+
+
+  public void delayBeforeSTOFlag(String message) {
+    if (isTestingSTO) {
+      //Only for testing
+      Times time = delays.get();
+      if (time == null){
+        time = saveTimes();
+      }
+      try {
+        LOG.debug("Testing STO. "+message+" Sleeping for " + time.delayBeforeSTOFlag);
+        Thread.sleep(time.delayBeforeSTOFlag);
+        LOG.debug("Testing STO. "+message+" Waking up from sleep of " + time.delayBeforeSTOFlag);
+      } catch (InterruptedException e) {
+        LOG.warn(e);
+      }
+    }
+  }
+
+  public void delayAfterBbuildingTree(String message) {
+    //Only for testing
+    if (isTestingSTO) {
+      Times time = delays.get();
+      if (time == null){
+        time = saveTimes();
+      }
+      try {
+        LOG.debug("Testing STO. "+message+" Sleeping for " + time.delayAfterBuildingTree);
+        Thread.sleep(time.delayAfterBuildingTree);
+        LOG.debug("Testing STO. "+message+" Waking up from sleep of " + time.delayAfterBuildingTree);
+      } catch (InterruptedException e) {
+        LOG.warn(e);
+      }
+    }
+  }
+
+  private class Times {
+    long delayBeforeSTOFlag = 0; //This parameter can not be more than TxInactiveTimeout: 1.2 sec
+    long delayAfterBuildingTree = 0;
+
+    public Times(long delayBeforeSTOFlag, long delayAfterBuildingTree) {
+      this.delayBeforeSTOFlag = delayBeforeSTOFlag;
+      this.delayAfterBuildingTree = delayAfterBuildingTree;
+    }
+  }
+
+  public void setTestingSTO(boolean val) {
+    this.isTestingSTO = val ;
+  }
+
+  public boolean isTestingSTO() {
+    return isTestingSTO;
+  }
+
+  private Times saveTimes(){
+    if(isTestingSTO) {
+      delays.remove();
+      Times times = new Times(delayBeforeSTOFlag, delayAfterBuildingTree);
+      delays.set(times);
+      return times;
+    }
+    throw new UnsupportedOperationException("Operation only supported for testing");
```

Comment:
Save time is called in a lot of place without checking first if it is a test, this breaks a lot of tests.